(cap:op_matematicas)=
# Operações matemáticas e tipos

Em sua maneira mais simples, Python pode ser usado como uma calculadora que segue as mesmas convenções de ordem de operações com que está acostumado.

```{note}
Seção da documentação oficial relevante para este capítulo: https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex
```

## Operações matemáticas binárias básicas

Aqui estão as operações binárias (operam em dois números) básicas em Python. Preste atenção na presença do ponto decimal nas entradas e nas respostas.

**Soma: `+`**

In [1]:
2 + 4

6

In [2]:
3.7 + 4.9

8.600000000000001

In [3]:
3.7 + 5

8.7

**Subtração: `-`**

Quando este símbolo é utilizado sozinho com um número, indica que é negativo (-5)

In [4]:
2 - 4

-2

In [5]:
3.7 - 4.9

-1.2000000000000002

In [6]:
3.7 - 4

-0.2999999999999998

**Multiplicação: `*`**

In [7]:
2 * 4

8

In [8]:
3.7 * 4.9

18.130000000000003

In [9]:
3.7 * 5

18.5

**Divisão: `/`**

In [10]:
4 / 2

2.0

In [11]:
3.7 / 4.9

0.7551020408163265

In [12]:
3.7 / 5

0.74

**Divisão inteira: `//`**

In [13]:
4 // 2

2

In [14]:
3.7 // 4.9

0.0

In [15]:
3.7 // 2

1.0

**Potenciação: `**`**

In [16]:
2 ** 4

16

In [17]:
3.7 ** 4.9

608.3989388909348

In [18]:
4 ** 0.5

2.0

**Operação módulo: `%`**

Retorna o resto de uma divisão inteira. Por exemplo

10 // 4 = 2 * 4 + 2, então 10 % 4 = 2

In [19]:
2 % 4

2

In [20]:
4 % 2

0

In [21]:
10 % 4

2

In [22]:
3.7 % 2

1.7000000000000002

Quando o símbolo `-` é utilizado sozinho perto de um número (operação unária), indica que é um número negativo. Existe o equivalente para números positivos, `+`, mas não é muito utilizado.

In [23]:
-4 - -7

3

In [24]:
-4 - +7

-11

In [25]:
3 ** (-1)

0.3333333333333333

In [26]:
3 ** -1

0.3333333333333333

O operador módulo se comporta um pouco diferente de outras linguagens quando se trata de números negativos. Leia o conteúdo [aqui](https://stackoverflow.com/questions/3883004/how-does-the-modulo-operator-work-on-negative-numbers-in-python) para entender melhor, se desejar. Em suma, o resultado é sempre positivo se o número à direita for positivo.

In [27]:
-10 % 4

2

In [28]:
10 % -4

-2

Parênteses podem ser adicionados por clareza e para mudar a ordem de operações.

In [29]:
-4 - (-2)

-2

In [30]:
-4 - (-2) ** 2

-8

In [31]:
(-4 - (-2)) ** 2

4

In [32]:
27 ** (1/3)

3.0

 De forma simplificada, a ordem de operações matemáticas está a seguir, da maior para menor precedência:

1. Operações dentro de parênteses.
2. Potenciação. Porém, quando se refere imediatamente a um número com o operador negativo, este toma precedência. Vide exemplo acima, `3 ** -1` é igual a `3 ** (-1)`
3. Operador negativo e positivo.
4. Multiplicação, divisão, divisão inteira, módulo
5. Adição e subtração
6. Deslocamentos de bits (explicados em uma [seção extra](extra:bitwise))
7. Operadores bit-a-bit (explicados em uma [seção extra](extra:bitwise))
8. Comparações

Mais informações podem ser encontradas [na documentação oficial aqui](https://docs.python.org/3/reference/expressions.html#operator-precedence)]

Existe mais um operador que não foi abordado aqui, e será abordado no futuro. É a multiplicação matricial, realizado com `@`, p.e. `mat1 @ mat2`.

## Dois tipos numéricos: `int` e `float`

Note que nos resultados dos anteriores, ocasionalmente ocorre o aparecimento de pontos ou não nos valores de entrada e nos resultados. Quando um número não possui um ponto de separação decimal, é um número inteiro `int` e quando tem o ponto, é um `float`, um número de ponto flutuante. No cotidiano, não classificamos os números dessa maneira. Se nos perguntam "quanto é 10 dividido por 4", dizemos "2,5" e não nos importamos com o fato de termos passado do domínio dos inteiros $\mathbb Z$ para os reais $\mathbb R$. Porém, dentro do computador, esses números são representados de uma maneira muito distinta. Quando algumas operações são feitas com eles, o Python realiza uma conversão quando necessário, para seguir com nosso entendimento intuitivo.

No caso da divisão (internamente chamada de `true division`, o resultado é sempre um `float`. Porém, em certas situações[^0], queremos uma divisão somente da parte inteira, logo o operador `//` existe.

Existem três maneiras de representar números em Python básico, os números inteiros `int` (1, 5, 10, 1000, -500)[^1] e os números de ponto flutuante `float` (1.1, 1000.3, 1E-7, etc.)[^2] e os números complexos, que serão abordados em breve. Para criar um `int`, basta não colocar um ponto no número, e para criar um `float`, basta colocar um (e só um) ponto.

Aqui, `int` e `float` são os *tipos* desses números, que por sua vez são *objetos*. Um *tipo* é uma descrição, um conjunto de regras que ditam como interpretar um trecho de memória e como interagir com outros tipos, e um *objeto* como a concretização dessa descrição. Em breve, notaremos que se não há nenhuma regra sobre como dois tipos devem interagir, aparecerá uma mensagem de erro.

```{note}
Em Python, tudo é um objeto, e todo objeto possui um tipo.
```

Como um exemplo simplista, imagine uma bancada cheia de abacates. Eles todos diferem, mas você sabe que são todos abacates, pois possui um modelo mental do que um abacate é (p.e. casca verde-escura, polpa verde-amarelada gordurosa com sabor adocicado e uma semente grande no centro). Aqui, o *conceito* de abacate é seu *tipo*, e as frutas em si são *objetos*.

Um computador temos que representar todo tipo de dado com zeros e uns (bits), a diferenciação de blocos de memória (sequências de bits) é feita somente pelo tipo associado ao bloco. Um mesmo bloco pode significar números, texto, imagens, música, etc., dependendo do contexto. Por exemplo, a sequência de bits '1100001' pode significar tanto o número 97 quando o caracter 'a'. Já o `float` "97." é representado pela seguinte sequência, gigante, pois o tamanho dos `float`s em Python é 64 bits: `0100000001011000010000000000000000000000000000000000000000000000`.

[^0]: Por exemplo, se você quer saber quantas laranjas cabem em uma caixa, utilizar a divisão `/` não irá retornar um valor satisfatório. Não cabem 6,3 laranjas, somente 6.

[^1]: Em Python, `int`s são representados como sequências de bits sem um limite teórico, somente prático (memória de seu computador). Isso é diferente de outras línguas que possuem um limite de, por exemplo, 32 ou 64 bits nos números inteiros, possibilitando representar $2^{32}$ ou $2^{64}$ valores diferentes.

[^2]: A representação e operações com `float`s estão descritas na norma [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754).


## Maneiras alternativas de representar números

Podemos adicionar *underlines* a números inteiros para facilitar nosso entendimento. Por exemplo, o 1 milhão pode ser representado como

In [33]:
1_000_000

1000000

E a posição dos *underlines* não precisa ser regular, mas não podem ser consecutivos. É relativamente comum, em softwares financeiros, utilizar números inteiros da menor unidade monetária ao invés de `float`s, por razões de precisão. Logo, R$123,45 é representado como 12345 centavos, ou

In [34]:
123_45

12345

`float`s também podem receber *underlines* para facilitar o entendimento. Diferente de `int`s, `float`s podem ser escritos em notação científica abreviada, da seguinte maneira:

In [35]:
2.34E-5 * 9.23E4 / 10_000E-2

0.021598199999999998

Vimos que para criar um número inteiro, somente precisamos escrevê-lo como estamos habituados, na base 10. Existem outras maneiras de representar números inteiros em bases diferentes. Em Python, conseguimos representá-los nas bases binária (2), octal (8) e hexadecimal (16). Para isso, adicionamos um prefixo. Para representar o número 14 (base 10), podemos fazê-lo das seguintes maneiras:

* Binário: `0b1110`, pois $1 \times 2^3 + 1 \times 2^2 + 1 \times 2^1 + 0 \times 2^0 = 14$
* Octal: `0o16`, pois $1 \times 8^1 + 6 \times 8^0 = 14$
* Hexadecimal: `0xe`, pois $14 \times 16^0 = 14$[^3]

[^3]: Na base hexadecimal, A=10, B=11, C=12, D=13, E=14, F=15. As letras podem ser maiúsculas ou não.

Para comprovar:

In [36]:
0b1110

14

In [37]:
0o16

14

In [38]:
0xe

14

É importante entender que tal representação é para *nosso benefício*, não do computador. Para o computador, tudo é armazenado da maneira binária.

## Limitações de números de ponto flutuante

Note que em vários exemplos mostrados acima, algumas operações pareciam fornecer valores errados. Por exemplo, `3.7 * 4.9` deveria fornecer `18.13` mas fornece `18.130000000000003`. Isso ocorre porque a representação interna de `float`s utiliza base 2. Qualquer fração que utiliza como base um número diferente de 2 é um decimal que se repete (como 3 é para 10). Isso resulta num "bug" clássico, que possui até um [site dedicado a ele](https://0.30000000000000004.com/). Honestamente, isso não é extremamente relevante para o uso comum deste curso. Mas se você for utilizar `float`s em sistemas financeiros ou simulações numéricas, é bom saber que essa limitação existe. Em Python, o pacote [`decimal`](https://docs.python.org/3/library/decimal.html?highlight=decimal#module-decimal) fornece uma maneira mais precisa de lidar com números de ponto flutuante escritos em base decimal.

In [39]:
0.1 + 0.2

0.30000000000000004

## Números complexos `complex`

Para criar um número complexo, basta colocar o sufixo `j` logo depois de um `int` ou `float`. Certas operações com `int`s e `float`s podem produzir números complexos também. Todas as operações descritas anteriormente funcionam também com números complexos, exceto a divisão inteira `//` e módulo `%`. Se você tentar aplicar essa operação, um `TypeError` será lançado. Erros são a maneira que Python tem de informar que alguma coisa inesperada aconteceu. Neste caso, como há a definição dessas operações no tipo `complex`, então o erro lançado é um `TypeError`.

In [40]:
1j

1j

In [41]:
7j * ((6 + 2j) / 2j + (3 - 2j)**2)

(105+42j)

In [42]:
1.23 + 1j

(1.23+1j)

In [43]:
(-1)**(1/2)

(6.123233995736766e-17+1j)

In [44]:
1 // 1j

TypeError: unsupported operand type(s) for //: 'int' and 'complex'

In [45]:
1 % 1j

TypeError: unsupported operand type(s) for %: 'int' and 'complex'

In [46]:
1j % 1j

TypeError: unsupported operand type(s) for %: 'complex' and 'complex'

## Comparações e booleanos

## Exercícios

Cada exercício mostrado aqui possui uma célula associada com uma possível solução.

### Extendendo uma classe

```{note}
Este exercício requer conhecimento de **classes**, abordado [aqui](cap:oop).
```

Vimos que números complexos não implementam os operadores `%` e `//`. Extenda a classe `complex` com essa funcionalidade, mesmo que não faça sentido matemático.

In [ ]:
%load "./Soluções dos exercícios/complex_mod.py"